### PR TITLE
[cmd/opampsupervisor] Handle OpAMP connection settings

### LINKED ISSUE #21043: [cmd/opampsupervisor] Handle OpAMP connection settings messages
### Component(s)

_No response_

### Is your feature request related to a problem? Please describe.

[Per the OpAMP specification](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#connection-settings-management), OpAMP servers should be able to configure network connection configurations for connections that agents make. Since it is acting on behalf of the Collector as an OpAMP agent, the Supervisor should support these messages.

### Describe the solution you'd like

Handle the [ConnectionSettingsOffers](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#connectionsettingsoffers-message), in particular fields with the following message types:
1. [OpAMPConnectionSettings](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#opampconnectionsettings)
2. [TelemetryConnectionSettings](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#telemetryconnectionsettings)

### Describe alternatives you've considered

_No response_

### Additional context

An issue for handling [OtherConnectionSettings](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#otherconnectionsettings) messages will follow.

### PR DESCRIPTION
**Pull request by srikanthccv**
**Link to tracking Issue:** <Issue number if applicable>

Part of #21043; based on top of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29848 to add test

**Testing:** <Describe what testing was performed and which tests were added.>

Added integration test

### CODE REVIEW

**Comment by srikanthccv** [comment]
While debugging the test failure I found that wsclient `Stop` gets indefinitely blocked by `ReceiverLoop` unless some error occurs. I attempted to fix that here https://github.com/open-telemetry/opamp-go/pull/240. @evan-bradley @tigrannajaryan take a look when you get some time.

**Comment by haoqixu** [comment]
> While debugging the test failure I found that wsclient `Stop` gets indefinitely blocked by `ReceiverLoop` unless some error occurs (or conn close). I attempted to fix that here [open-telemetry/opamp-go#240](https://github.com/open-telemetry/opamp-go/pull/240). @evan-bradley @tigrannajaryan take a look when you get some time.

I think open-telemetry/opamp-go#213 will fix this. 🤔

**Comment by srikanthccv** [comment]
Both fixes should be there because `ReceiverLoop` should stop on the context cancellation as well.

**Comment by haoqixu** [comment]
> Both fixes should be there because `ReceiverLoop` should stop on the context cancellation as well.

It seems that `gorilla/websocket` doesn't provide a method to unblock the bloking reading of connection. And we need to close the connection to unblock the reading.

If we consider `ReceiverLoop` as an internal implementation detail of the `wsclient` and we could gracefully shut down the `wsclient`, I think it's not a problem.

**Comment by srikanthccv** [comment]
I see these two fixes as complementary. Even as an internal implementation detail, I believe the context cancellation needs to be respected and closing the connection shouldn't be the only way to break out the receiver loop.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (srikanthccv) [inline cmd/opampsupervisor/supervisor/supervisor.go:422]: This must be removed after the graceful shutdown gets merged. Wanted to get the reviews so skipping this temporarily.

**Comment by srikanthccv** [comment]
I believe this is ready for reviews.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:416]: Should we use this method in `Supervisor.Shutdown`?

File: cmd/opampsupervisor/e2e_test.go
Comment (evan-bradley) [inline cmd/opampsupervisor/e2e_test.go:136]: How noisy is this when `go test` is run without the `-v` flag?

**Review by evan-bradley — COMMENTED** [review]
Overall looks good to me.

File: cmd/opampsupervisor/e2e_test.go
Comment (srikanthccv) [reply cmd/opampsupervisor/e2e_test.go:136]: It was noisy but not too much. I can revert this change if you think the logs are too many. It was helpful for me to debug the failed tests.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:423]: Updated

**Review by evan-bradley — APPROVED** [review]
Looks good to me, thanks for adding this.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:483]: startOpAMP() initiates but does not wait for successful connection. I think we need to wait for it before we consider the new opamp settings successful. It can be done in a future PR, but would be good to add a TODO here.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:444]: Should we try to perform this operation in a separate goroutine so that we don't block here in this callback for the entire duration of switching the connections? Let the callback return quickly and initiate the re-establishing of the connection in a separate goroutine.

This would ensure that if the currently received message contains more data and not just connection settings all that data will be processed. With the current approach I am not sure what exactly will happen.

I think callbacks generally should avoid doing long-lasting blocking operations since they than block other callbacks and the entire opamp operation.

I also think [the comment here](https://github.com/open-telemetry/opamp-go/blob/7e92da0f17ef9f2fd0a387dd6b62b451c80f4207/client/types/callbacks.go#L72) is completely misleading. We should either implement what the comment says (i.e. the caller should do the reconnection) or fix the comment to sat the callback implementation should do the reconnection. @andykellr @evan-bradley Any thoughts on what you would prefer?

Comment (andykellr) [reply cmd/opampsupervisor/supervisor/supervisor.go:444]: Our agent does not currently support `AcceptsOpAMPConnectionSettings` so I have not spent much time with this part of the implementation. Looking at the code, the following things are confusing to me:

1. There are several references to `OnRemoteConfig` (e.g. "See OnRemoteConfig for the behavior."). This was removed a while ago.

2. This sentence doesn't make sense:

> 	// If OnOpampConnectionSettings returns nil and then the caller will
>	// attempt to reconnect to the OpAMP Server using the new settings.

We should clearly state who is responsible for establishing a new connection. I could imagine this being fully implemented in the client library and only have `OnOpampConnectionSettings` responsible for validating the fields (e.g. endpoint matching a list of acceptable servers or certificate signed by acceptable authority). If this returns nil, the library would attempt to start a new client with the new settings, send an initial status message, and stop and transition to the new connection on success.

If we expect the implementer of the callback to do this, we should describe the process we expect to take place and implement that in tests and the example agent.

3. If we do this in a separate goroutine like the example agent in the opamp-go library, we have no way of knowing if it will be successful and should be accepted. Returning nil from the callback will cause the settings to be accepted. If we create a separate goroutine we will have to wait for it before returning here.

https://github.com/open-telemetry/opamp-go/blob/7e92da0f17ef9f2fd0a387dd6b62b451c80f4207/client/internal/receivedprocessor.go#L198-L202

It is not possible to confirm success until the client is started and an initial status message is sent.

4. OnOpampConnectionSettingsAccepted references OnOpampConnectionSettingsOffer which was renamed OnOpampConnectionSettings

Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:444]: > This would ensure that if the currently received message contains more data and not just connection settings all that data will be processed. With the current approach I am not sure what exactly will happen.

`onOpampConnectionSettings` is the last part of the message processing in client implementation.

> If we do this in a separate goroutine like the example agent in the opamp-go library, we have no way of knowing if it will be successful and should be accepted.

Right, the returned value from this callback indicates the reject/accept status of the connection settings. I don't see how we can make this async without changing how the client is expected to handle `OnOpampConnectionSettings`. Also, we don't report the error as described in `OnOpampConnectionSettings` https://github.com/open-telemetry/opamp-spec/issues/164.

File: cmd/opampsupervisor/supervisor/config/config.go
Comment (MovieStoreGuy) [inline cmd/opampsupervisor/supervisor/config/config.go:21]: Sorry I had missed this in earlier, and isn't strictly relevant for the PR but I am curious, what does a pointer value add here?

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (MovieStoreGuy) [inline cmd/opampsupervisor/supervisor/supervisor.go:452]: I wonder if there any value in using some of the `confmap` work to merge these values together instead of needing to update each one explicitly?

**Review by MovieStoreGuy — COMMENTED** [review]
I don't have much to add here, I don't see any potential issues here and @tigrannajaryan has more context here.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/supervisor.go:444]: >We should clearly state who is responsible for establishing a new connection. I could imagine this being fully implemented in the client library and only have OnOpampConnectionSettings responsible for validating the fields (e.g. endpoint matching a list of acceptable servers or certificate signed by acceptable authority). If this returns nil, the library would attempt to start a new client with the new settings, send an initial status message, and stop and transition to the new connection on success.

I think this was the original intent. However, I am not sure this is the best approach. It is also not how the example implementation works and I don't see good arguments against how the example is implemented today.

To summarize this is what the example is supposed to do:

1. OnOpampConnectionSettings callback called when new settings are offered by the Server.
2. The callback implementation first pre-verifies the settings (e.g. check certificates, etc).
3. If checks in (2) pass the callback implementation logs that it is starting to reconnect, creates a reconnection goroutine and returns from the callback. 
4. Reconnection goroutine stops the Client, creates a new Client with new connection settings and wait for successful OnConnect() callback.
5. If OnConnect() is not called within a predefined period of time reconnection goroutine assumes the new connection settings are bad, reverts to the old connection settings and re-creates the Client again.
6. We get rid of OnOpampConnectionSettingsAccepted(), it is not needed.

The example implementation of steps 4 and 5 is not complete today (e.g not waiting for OnConnect), but can be modified to match what I described.

Thoughts?

Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/supervisor.go:444]: I created an issue for this https://github.com/open-telemetry/opamp-go/issues/261

Let's also discuss today in our call.

File: cmd/opampsupervisor/supervisor/config/config.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/config/config.go:21]: Since the zero-value for `bool` is `false` but the default value may be `true`, using a pointer instead allows us to use a nil pointer value as a way to represent when the user didn't pass a value in and a default should be used.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/supervisor.go:444]: FYI, https://github.com/open-telemetry/opamp-go/pull/266 removes OnOpampConnectionSettingsAccepted callback.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:452]: I am not very familiar with `confmap` work. I will take a look at it and If it helps here I can send a separate PR for it.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:451]: Update the implementation to run in a separate goroutine and use OnConnect to determine the connection status.

**Comment by MovieStoreGuy** [comment]
Could I ask you to clean up the mod file?

**Pull request merged**
